### PR TITLE
[AI POC] Expert finder js sample

### PR DESCRIPTION
**Pull request by Prithvi-MSFT**

### CODE REVIEW

File: samples/expert-finder-js/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+---
+page_type: sample
+description: This sample demonstrates how to use a copilot plugin that can be used to search candidates for job. This sample can also be used as a search based messaging extension.
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:3]: This sample name should be: msgext-expert-finder and it should be linked to Root readme file as well.

Comment (ChetanSharma-msft) [reply samples/expert-finder-js/README.md:3]: Summary should be something like below:

image

File: samples/expert-finder-js/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+products:
+- office-teams
+- copilot-m365
+languages:
+- javascript
+---
+
+# Expert Finder Copilot Plugin
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:11]: Please refer this: 
image
https://github.com/OfficeDev/Copilot-for-M365-Samples/tree/main/samples/msgext-northwind-inventory-action-ts

File: samples/expert-finder-js/README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+Please find below demo manifest which is deployed on Microsoft Azure and you can try it yourself by uploading the app package (.zip file link below) to your teams and/or as a personal app. (Sideloading must be enabled for your tenant, [see steps here](https://docs.microsoft.com/microsoftteams/platform/concepts/build-and-test/prepare-your-o365-tenant#enable-custom-teams-apps-and-turn-on-custom-app-uploading)).
+
+**Microsoft Teams Expert Finder sample app:** [Manifest](/samples/msgext-doc-compliance-checker/demo-manifest/expert-finder.zip)
+
+## Prerequisites
+
+- [Node.js 18.x](https://nodejs.org/download/release/v18.18.2/)
+- [Visual Studio Code](https://code.visualstudio.com/)
+- [Blob Storage](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-portal)
+- [Teams Toolkit](https://marketplace.visualstudio.com/items?itemName=TeamsDevApp.ms-teams-vscode-extension)
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:27]: image

File: samples/expert-finder-js/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+- javascript
+---
+
+# Expert Finder Copilot Plugin
+
+## Interaction with app
+
+ ![expert-finder-gif](images/expert-finder.gif)
+
+## Try it yourself - experience the App in your Microsoft Teams client
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:17]: Will it work directly without creating table storage?

File: samples/expert-finder-js/README.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+Please find below demo manifest which is deployed on Microsoft Azure and you can try it yourself by uploading the app package (.zip file link below) to your teams and/or as a personal app. (Sideloading must be enabled for your tenant, [see steps here](https://docs.microsoft.com/microsoftteams/platform/concepts/build-and-test/prepare-your-o365-tenant#enable-custom-teams-apps-and-turn-on-custom-app-uploading)).
+
+**Microsoft Teams Expert Finder sample app:** [Manifest](/samples/msgext-doc-compliance-checker/demo-manifest/expert-finder.zip)
+
+## Prerequisites
+
+- [Node.js 18.x](https://nodejs.org/download/release/v18.18.2/)
+- [Visual Studio Code](https://code.visualstudio.com/)
+- [Blob Storage](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-portal)
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:26]: Where are you using blob storage ?

File: samples/expert-finder-js/README.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+The simplest way to run this sample in Teams is to use Teams Toolkit for Visual Studio Code.
+
+1. Ensure you have downloaded and installed [Visual Studio Code](https://code.visualstudio.com/docs/setup/setup-overview)
+1. Install the [Teams Toolkit extension](https://marketplace.visualstudio.com/items?itemName=TeamsDevApp.ms-teams-vscode-extension)
+1. Select **File > Open Folder** in VS Code and choose this samples directory from the repo
+1. Using the extension, sign in with your Microsoft 365 account where you have permissions to upload custom apps
+1. Select **Debug > Start Debugging** or **F5** to run the app in a Teams web client.
+1. In the browser that launches, select the **Add** button to install the app to Teams.
+
+> If you do not have permission to upload custom apps (sideloading), Teams Toolkit will recommend creating and using a Microsoft 365 Developer Program account - a free program to get your own dev environment sandbox that includes Teams.
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:40]: Add doc link

File: samples/expert-finder-js/README.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+3) Follow schema as given in below image while creating azure table.
+   ![Schema](Images/table-storage-schema.png)
+
+4) Clone the repository
+
+    ```bash
+    git clone https://github.com/OfficeDev/Microsoft-Teams-Samples.git
+    ```
+5) Navigate to the `samples/expert-finder-js` folder and open with Visual Studio Code.
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:57]: Update these paths as per new folder.

File: samples/expert-finder-js/README.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+6) Navigate to the `samples/expert-finder-js/.localConfigs` directory and update the values below.
+
+   ```txt
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:61]: Steps are missing to run this sample and upload the manifest.

File: samples/expert-finder-js/README.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+      CONNECTION_STRING=<table storage connection string> (Created in step 2)
+      TABLE_NAME=<table name> (Created in step 2)
+   ```
+
+### Test the app in Copilot for Microsoft 365
+Navigate to the Microsoft Copilot for Microsoft 365 chat. Check the lower left of the chat user interface, below the compose box. You should see a plugin icon. Click this and enable theExpert-finder plugin.
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:67]: Add screenshot to see plugin icon.

File: samples/expert-finder-js/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+- Use a basic prompt: `Find candidates with skill in Azure.`
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:69]: Add 3-4 prompts to test.

File: samples/expert-finder-js/README.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+> If you do not have permission to upload custom apps (sideloading), Teams Toolkit will recommend creating and using a Microsoft 365 Developer Program account - a free program to get your own dev environment sandbox that includes Teams.
+
+## Setup and use the sample
+
+1) Create azure storage account
+   Refer the document [Create storage account]("https://learn.microsoft.com/en-us/azure/storage/common/storage-account-create?tabs=azure-portal") and create a storage account in azure.
+
+2) Follow the steps mentioned to [Create azure table storage]("https://learn.microsoft.com/en-us/azure/storage/tables/table-storage-quickstart-portal") inside storage account.
+
+3) Follow schema as given in below image while creating azure table.
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/README.md:49]: Mention the step to update the values in table storage.

File: samples/expert-finder-js/assets/sample.json [hunk 12]
```diff
@@ -0,0 +1,60 @@
+[
+  {
+    "name": "officedev-copilot-for-m365-plugins-samples-msgext-multiparam-js",
+    "source": "officedev",
+    "title": "Using multiple parameters in a plugin for Microsoft Copilot for Microsoft 365 using JavaScript",
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/assets/sample.json:5]: **This file is totally wrong.**

File: samples/expert-finder-js/infra/azure.bicep [hunk 13]
```diff
@@ -0,0 +1,82 @@
+@maxLength(20)
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/infra/azure.bicep:1]: .png images are not complete Teams UI like we do for other samples.

File: samples/expert-finder-js/src/searchApp.js [hunk 14]
```diff
@@ -0,0 +1,268 @@
+const { TeamsActivityHandler, CardFactory } = require("botbuilder");
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/src/searchApp.js:1]: File header is missing in all js files.

File: samples/expert-finder-js/src/searchApp.js [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+    // Create a table service object using the connection string
+    const tableService = azure.createTableService(connectionString);
+
+    var candiDateData = [];
+
+    // Define the name of the table you want to store data in
+    const tableName = config.tableName;
+
+    // When the Bot Service Auth flow completes, the query.State will contain a magic code used for verification.
+    const magicCode =
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/src/searchApp.js:58]: What is this magic code?

File: samples/expert-finder-js/src/searchApp.js [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+                }
+                whereClause += `(${availabilityCondition})`;
+            }
+
+            // If no parameters provided, select all
+            if (whereClause === "") {
+              whereClause = "PartitionKey ne ''"; // Dummy condition to select all in case parameters are null or empty
+          }
+    
+            query.where(whereClause);
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/src/searchApp.js:130]: Add proper spacing

File: samples/expert-finder-js/src/searchApp.js [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+            }
+
+            resolve(filteredCandidates);
+          }
+        );
+      });
+    }
+
+    var candidates = await fetchCandidates(searchObject);
+
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/src/searchApp.js:165]: Line spacing check all the places.

File: samples/expert-finder-js/src/searchApp.js [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+          },
+          {
+            "type": "ColumnSet",
+            "columns": [
+              {
+                "type": "Column",
+                "items": [
+                  {
+                    "type": "Image",
+                    "url": "https://pbs.twimg.com/profile_images/3647943215/d7f12830b3c17a5a9e4afcc370e3a37e_400x400.jpeg",
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/src/searchApp.js:197]: What is this hardcoded image?

File: samples/expert-finder-js/teamsapp.local.yml [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+  - uses: cli/runNpmCommand
+    name: install dependencies
+    with:
+      args: install --no-audit
+  # Generate runtime environment variables
+  # - uses: file/createOrUpdateEnvironmentFile
+  #   with:
+  #     target: ./.localConfigs
+  #     envs:
+  #       BOT_ID: ${{BOT_ID}}
```
Comment (ChetanSharma-msft) [inline samples/expert-finder-js/teamsapp.local.yml:88]: I think, we should add connecting string and table name here as well?

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added my review comments.

File: samples/expert-finder-js/README.md (hunk 3 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:27]: added

File: samples/expert-finder-js/README.md (hunk 5 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:26]: removed

File: samples/expert-finder-js/README.md (hunk 7 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:57]: updated

File: samples/expert-finder-js/README.md (hunk 11 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:49]: added as part of step 1

File: samples/expert-finder-js/README.md (hunk 4 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:17]: added steps for creating storage account

File: samples/expert-finder-js/README.md (hunk 8 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:61]: The test the app in copilot mentions the steps

File: samples/expert-finder-js/assets/sample.json (hunk 12 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/assets/sample.json:5]: changed

File: samples/expert-finder-js/README.md (hunk 2 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:11]: changed

File: samples/expert-finder-js/README.md (hunk 1 above)
Comment (Prithvi-MSFT) [reply samples/expert-finder-js/README.md:3]: changed

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added my review comments.

File: samples/msgext-expert-finder-js/.gitignore [hunk 20]
```diff
@@ -0,0 +1,11 @@
+# TeamsFx files
+.localConfigs
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/.gitignore:2]: added

File: samples/msgext-expert-finder-js/appPackage/manifest.json [hunk 21]
```diff
@@ -0,0 +1,76 @@
+{
+  "$schema": "https://developer.microsoft.com/json-schemas/teams/v1.17/MicrosoftTeams.schema.json",
+  "manifestVersion": "1.17",
+  "version": "1.0.0",
+  "id": "${{TEAMS_APP_ID}}",
+  "localizationInfo": {
+    "defaultLanguageTag": "en-us"
+  },
+  "developer": {
+    "name": "Teams App, Inc.",
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/appPackage/manifest.json:10]: changed

File: samples/msgext-expert-finder-js/README.md [hunk 22]
```diff
@@ -0,0 +1,185 @@
+---
+page_type: sample
+description:  This sample implements a Teams message extension that can be used to query a database of candidates based on their skills, location and availability. The sample, when used with Copilot for M365, demonstrates SSO, and Copilot's ability to perform multi-parameter search.
+products:
+- office-teams
+- copilot-m365
+languages:
+- javascript
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/README.md:8]: added

File: samples/msgext-expert-finder-js/assets/sample.json [hunk 23]
```diff
@@ -0,0 +1,68 @@
+[
+  {
+    "name": "officedev-microsoft-teams-samples-msgext-expert-finder-js",
+    "source": "officeDev",
+    "title": "Microsoft Teams msgext-expert-finder javascript sample",
+    "shortDescription": "Microsoft Teams bot-ai-meeting-helper sample app in Node.js how Azure Open AI extracts action items from meeting transcriptions for all participants who subscribed to the meeting, and subsequently sends these action items to each individual user in a 1:1 chat after the meeting concludes.",
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/assets/sample.json:6]: changed

File: samples/msgext-expert-finder-js/assets/sample.json [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    "url": "https://github.com/OfficeDev/Microsoft-Teams-Samples/tree/main/samples/bot-ai-meeting-helper",
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/assets/sample.json:7]: changed

File: samples/msgext-expert-finder-js/package.json [hunk 25]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "msgextmultiparamjs",
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/package.json:2]: changed

File: samples/msgext-expert-finder-js/package.json [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+  "version": "1.0.0",
+  "msteams": {
+    "teamsAppId": null
+  },
+  "description": "Microsoft Teams Toolkit message extension search sample",
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/package.json:7]: changed

File: samples/msgext-expert-finder-js/src/searchApp.js [hunk 27]
```diff
@@ -0,0 +1,268 @@
...
+class SearchApp extends TeamsActivityHandler {
+
+  async handleTeamsMessagingExtensionQuery(context, query) {
+    const { parameters } = query;
+
+    const skills = getParameterByName(parameters, "Skill");
+    const country = getParameterByName(parameters, "Location");
+    const availabilityParam = getParameterByName(parameters, "Availability");
+
+    var availability;
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/searchApp.js:14]: based on db schema the name is given

File: samples/msgext-expert-finder-js/src/searchApp.js [hunk 28]
```diff
@@ -0,0 +1,268 @@
...
+        filterObject.availability = availability;
+      }
+
+      return filterObject;
+    }
+
+    const searchObject = constructSearchObject(skills, country, availability);
+
+    // Define your Azure Table Storage connection string or credentials
+    const connectionString = config.connectionString;
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/searchApp.js:47]: changed

File: samples/msgext-expert-finder-js/src/searchApp.js [hunk 29]
```diff
@@ -0,0 +1,268 @@
...
+
+    // Create a table service object using the connection string
+    const tableService = azure.createTableService(connectionString);
+
+    var candiDateData = [];
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/searchApp.js:52]: changed

File: samples/msgext-expert-finder-js/src/searchApp.js [hunk 30]
```diff
@@ -0,0 +1,268 @@
...
+        : '';
+
+    const tokenResponse = await context.adapter.getUserToken(
+      context,
+      "authbot",
+      magicCode
+    );
+
+    if (!tokenResponse || !tokenResponse.token) {
+      //     // There is no token, so the user has not signed in yet.
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/searchApp.js:70]: removed

File: samples/msgext-expert-finder-js/src/searchApp.js [hunk 31]
```diff
@@ -0,0 +1,268 @@
...
+          whereClause += `(${availabilityCondition})`;
+        }
+
+        // If no parameters provided, select all
+        if (whereClause === "") {
+          whereClause = "PartitionKey ne ''"; // Dummy condition to select all in case parameters are null or empty
+        }
+
+        query.where(whereClause);
+
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/searchApp.js:131]: removed

File: samples/msgext-expert-finder-js/src/searchApp.js [hunk 32]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    // Fetch candidates based on applied filters.
+    var candidates = await fetchCandidates(searchObject);
+
+    var attachments = [];
+    candiDateData = candidates;
+    console.log("Candidates:", candiDateData);
+
+
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/searchApp.js:171]: removed

File: samples/msgext-expert-finder-js/README.md [hunk 33]
```diff
@@ -0,0 +1,185 @@
...
+---
+
+# Expert Finder message extension sample with SSO
+
+## Interaction with app
+
+In this sample, we demonstrate two capabilities of Copilot for M365:
+1) to search across a database of candidates to find a relevant "expert" based on multiple parameters such as their skill, location and availability.
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/README.md:16]: changed

File: samples/msgext-expert-finder-js/README.md [hunk 34]
```diff
@@ -0,0 +1,185 @@
...
+On first use, a window requesting for Sign In pops-up and once SSO is completed, the user flow continues as usual. 
+
+The database of candidates is hosted in an Azure Table storage, and the app itself is a messaging extension that can be used as a plug-in within Copilot for M365.
+
+ ![expert-finder-gif](images/expert-finder.gif)
+
+## Try it yourself - experience the App in your Microsoft Teams client
+Please find below demo manifest which is deployed on Microsoft Azure and you can try it yourself by uploading the app package (.zip file link below) to your teams and/or as a personal app. (Sideloading must be enabled for your tenant, [see steps here](https://docs.microsoft.com/microsoftteams/platform/concepts/build-and-test/prepare-your-o365-tenant#enable-custom-teams-apps-and-turn-on-custom-app-uploading)).
+
+**Microsoft Teams Expert Finder sample app:** [Manifest](/samples/msgext-epert-finder-js/demo-manifest/expert-finder.zip)
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/README.md:28]: changed

File: samples/msgext-expert-finder-js/README.md [hunk 35]
```diff
@@ -0,0 +1,185 @@
...
+
+## Prerequisites
+
+- [Node.js 18.x](https://nodejs.org/download/release/v18.18.2/)
+- [Visual Studio Code](https://code.visualstudio.com/)
+- [Blob Storage](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-portal)
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/README.md:34]: changed

File: samples/msgext-expert-finder-js/README.md [hunk 36]
```diff
@@ -0,0 +1,185 @@
...
+![Plugin](images/plugin.png)
+
+- Here are some sample prompts you can try : 
+1) `Find candidates with skill in Azure.`
+2) `Find candidates with skill in React and who are from India.`
+3) `Find candidates with skill in Azure and available for interview.`
+
+## Running the sample
+
+**SSO auth**
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/README.md:168]: added

File: samples/msgext-expert-finder-js/README.md [hunk 37]
```diff
@@ -0,0 +1,185 @@
+---
+page_type: sample
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/README.md:2]: linked

File: samples/msgext-expert-finder-js/src/config.js [hunk 38]
```diff
@@ -0,0 +1,8 @@
+const config = {
+  botId: process.env.BOT_ID,
+  botPassword: process.env.BOT_PASSWORD,
+  connectionString: process.env.CONNECTION_STRING,
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/config.js:4]: changed

File: samples/msgext-expert-finder-js/src/searchApp.js [hunk 39]
```diff
@@ -0,0 +1,268 @@
+const { TeamsActivityHandler, CardFactory } = require("botbuilder");
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/src/searchApp.js:1]: added

File: samples/msgext-expert-finder-js/README.md [hunk 40]
```diff
@@ -0,0 +1,185 @@
...
+
+    ```bash
+    git clone https://github.com/OfficeDev/Microsoft-Teams-Samples.git
+    ```
+6) Navigate to the `samples/msgext-expert-finder-js` folder and open with Visual Studio Code.
+
+7) Navigate to the `samples/msgext-expert-finder-js/.localConfigs` directory and update the values below.
+
+   ```txt
+      CONNECTION_STRING=<azure storage account connection string> (Created in step 1)
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/msgext-expert-finder-js/README.md:58]: updated

**Review by Wajeed-msft — APPROVED** [review]

**Review by ChetanSharma-msft — APPROVED** [review]
Changes looks fine, approving it.

**Pull request merged**